### ISSUE TITLE
[java]

### ISSUE THREAD

**Issue by barriosnahuel** [pmd/pmd#2011]
**Affects PMD Version:** v6.13.0

**Rule:** UnusedFormalParameter

**Description:** False positive on Android projects when defining a method with an unused parameter. This parameter is required in order to let the OS call this method directly from a layout for instance after a user clicks on a button.

**Code Sample demonstrating the issue:**

```
public void saveButton(final View view) {
    return;
}
```

**Running PMD through:** *Other* (https://github.com/Monits/static-code-analysis-plugin v2.6.10)

**Comment by jsotuyod** [comment]
@barriosnahuel I assume, this is a listener set directly from the XML using the `onclick` attribute…

PMD has no way to know that argument is needed by convention (as it's not required by an explicit contract), so it can't whitelist it. Moreover, PMD can't if the method is being used at all, as there may even be no explicit calls to the method anywhere in the codebase.

If using a clicklistener, this method should be an `@Override` and not be flagged. This approach should be preferred also as it reduces coupling between the activity and the layout (the activity needs to know the layout, but not the other way around), and is way easier to refactor safely / without runtime errors (specially if using view binding).

**Comment by barriosnahuel** [comment]
Thanks! That's the use case I say.

I understand the click listener approach is better, but what I said is still a feature of the Android platform. Based on this argument, is there any plans on PMD for instance to add this kind of features as an extension for Android projects perhaps?

**Comment by jsotuyod** [comment]
@barriosnahuel doing so would require both multifile and multi-language analysis. Both are in our roadmap, but not going to happen in the short term.

In the meantime, you can either suppress this locally, generally with an XPath expression in the rule config, or disable the `checkAll` flag from the rule to skip all non-private methods altogether (would probably introduce false negatives instead).
